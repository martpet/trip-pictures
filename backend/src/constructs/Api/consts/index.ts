export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
