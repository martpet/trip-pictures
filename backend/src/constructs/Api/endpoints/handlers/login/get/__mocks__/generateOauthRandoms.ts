export const generateOauthRandoms = () => ({
  stateNonce: 'dummyStateNonce',
  codeChallenge: 'dummyCodeChallenge',
  codeVerifier: 'dummyCodeVerifier',
});
