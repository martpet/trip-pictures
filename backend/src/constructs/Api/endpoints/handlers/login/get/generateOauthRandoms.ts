import { crypto, getRandomBase64UrlSafe } from 'lambda-layer';

export async function generateOauthRandoms() {
  const idTokenNonce = await getRandomBase64UrlSafe(128);
  const stateNonce = await getRandomBase64UrlSafe(128);
  const codeVerifier = await getRandomBase64UrlSafe(128);

  const codeChallenge = crypto
    .createHash('sha256')
    .update(codeVerifier)
    .digest('base64url');

  return {
    idTokenNonce,
    stateNonce,
    codeVerifier,
    codeChallenge,
  };
}
