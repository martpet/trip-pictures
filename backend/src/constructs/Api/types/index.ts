export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
