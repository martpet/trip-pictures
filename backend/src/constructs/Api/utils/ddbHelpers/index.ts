export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
