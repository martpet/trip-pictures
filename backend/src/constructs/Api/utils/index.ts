export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
