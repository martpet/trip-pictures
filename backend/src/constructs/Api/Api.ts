import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
