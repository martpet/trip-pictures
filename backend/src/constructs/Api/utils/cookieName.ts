import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
