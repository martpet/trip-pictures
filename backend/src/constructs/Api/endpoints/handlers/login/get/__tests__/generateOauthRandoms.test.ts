import crypto from 'crypto';

import { itCalls, itResolves } from '~/constructs/Api/utils';
import { getRandomBase64UrlSafe } from '~/utils';

import { generateOauthRandoms } from '../generateOauthRandoms';

vi.mock('crypto');
vi.mock('~/utils/getRandomBase64UrlSafe');

beforeEach(() => {
  vi.mocked(getRandomBase64UrlSafe)
    .mockResolvedValueOnce('dummyIdTokenNonce')
    .mockResolvedValueOnce('dummyStateNonce')
    .mockResolvedValueOnce('dummyCodeVerifier');
});

describe('generateOauthRandoms', () => {
  itResolves(generateOauthRandoms);
  itCalls(getRandomBase64UrlSafe, generateOauthRandoms);
  itCalls(crypto.createHash, generateOauthRandoms);
  itCalls(crypto.createHash('').update, generateOauthRandoms);
  itCalls(crypto.createHash('').update('').digest, generateOauthRandoms);
});
