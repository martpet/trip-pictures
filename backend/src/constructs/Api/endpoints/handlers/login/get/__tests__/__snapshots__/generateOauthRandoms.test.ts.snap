// Vitest Snapshot v1

exports[`generateOauthRandoms > calls "createHash" with correct args 1`] = `
[
  [
    "sha256",
  ],
]
`;

exports[`generateOauthRandoms > calls "digest" with correct args 1`] = `
[
  [
    "base64url",
  ],
]
`;

exports[`generateOauthRandoms > calls "getRandomBase64UrlSafe" with correct args 1`] = `
[
  [
    128,
  ],
  [
    128,
  ],
  [
    128,
  ],
]
`;

exports[`generateOauthRandoms > calls "update" with correct args 1`] = `
[
  [
    "dummyCodeVerifier",
  ],
]
`;

exports[`generateOauthRandoms > resolves with a correct value 1`] = `
{
  "codeChallenge": "dummyCryptoCreateHashUpdateDigest",
  "codeVerifier": "dummyCodeVerifier",
  "idTokenNonce": "dummyIdTokenNonce",
  "stateNonce": "dummyStateNonce",
}
`;
